In [30]:
!pip install ydata-profiling
!pip install --upgrade Pillow

  Using cached ydata_profiling-4.9.0-py2.py3-none-any.whl (356 kB)
  Using cached numba-0.60.0-cp39-cp39-win_amd64.whl (2.7 MB)
  Using cached pydantic-2.8.2-py3-none-any.whl (423 kB)
  Using cached htmlmin-0.1.12-py3-none-any.whl
  Using cached typeguard-4.3.0-py3-none-any.whl (35 kB)
  Using cached ImageHash-4.3.1-py2.py3-none-any.whl (296 kB)
  Using cached wordcloud-1.9.3-cp39-cp39-win_amd64.whl (300 kB)
  Using cached phik-0.12.4-cp39-cp39-win_amd64.whl (666 kB)
  Using cached dacite-1.8.1-py3-none-any.whl (14 kB)
  Using cached llvmlite-0.43.0-cp39-cp39-win_amd64.whl (28.1 MB)
  Using cached annotated_types-0.7.0-py3-none-any.whl (13 kB)

ERROR: Cannot uninstall 'llvmlite'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.



  Attempting uninstall: llvmlite
    Found existing installation: llvmlite 0.38.0


In [1]:
import numpy as np
import pandas as pd
from ydata_profiling import ProfileReport
import os as os

ModuleNotFoundError: No module named 'ydata_profiling'

# Read cleaned dataset

In this section, we read the cleaned dataset, and check for duplicates and missing values:

In [2]:
df = pd.read_csv("CARES_data_cleaned.csv")
df

Indexno   Age  Gender  RCRI score Anemia category  PreopEGFRMDRD  \
0            2  48.0  Female         NaN             NaN            NaN   
1            5  36.0  Female         NaN             NaN            NaN   
2            6  64.0  Female         NaN            Mild     152.538570   
3            9  73.0    Male         NaN        Moderate     117.231496   
4           10  73.0    Male         0.0            Mild      98.651255   
...        ...   ...     ...         ...             ...            ...   
90780   121192  66.0  Female         2.0            Mild      48.339582   
90781   121194  50.0    Male         1.0        Moderate     126.592489   
90782   121197  58.0  Female         NaN             NaN      86.306771   
90783   121198  63.0  Female         0.0             NaN            NaN   
90784   121200  45.0  Female         0.0            Mild     125.902498   

      GradeofKidneydisease  DaysbetweenDeathandoperation  @30daymortality  \
0                      NaN                           NaN            False   
1                      NaN                           NaN            False   
2                       G1                           NaN            False   
3                       G1                           NaN            False   
4                       G1                          59.0            False   
...                    ...                           ...              ...   
90780                  G3a                           NaN            False   
90781                   G1                           NaN            False   
90782                   G2                           NaN            False   
90783                  NaN                           NaN            False   
90784                   G1                           NaN            False   

       Preoptransfusionwithin30days  ...  \
0                               0.0  ...   
1                               0.0  ...   
2                               0.0  ...   
3                               0.0  ...   
4                               0.0  ...   
...                             ...  ...   
90780                           0.0  ...   
90781                           0.0  ...   
90782                           0.0  ...   
90783                           0.0  ...   
90784                           0.0  ...   

       Anemiacategorybinned_Moderate/Severe  Anemiacategorybinned_NA  \
0                                     False                     True   
1                                     False                     True   
2                                     False                    False   
3                                      True                    False   
4                                     False                    False   
...                                     ...                      ...   
90780                                 False                    False   
90781                                  True                    False   
90782                                 False                     True   
90783                                 False                     True   
90784                                 False                    False   

       RDW15.7_<= 15.7 RDW15.7_>15.7 RDW15.7_NA ASAcategorybinned_I  \
0                False         False       True                True   
1                 True         False      False                True   
2                 True         False      False                True   
3                 True         False      False                True   
4                False          True      False               False   
...                ...           ...        ...                 ...   
90780             True         False      False               False   
90781             True         False      False               False   
90782             True         False      False               False   
90783            False         False       True               False   
90784 

In [4]:
df.describe()

Indexno           Age    RCRI score  PreopEGFRMDRD  \
count   90785.000000  90785.000000  63361.000000   79955.000000   
mean    60334.001057     52.253225      0.322296      96.426155   
std     34930.467992     17.087307      0.634789      33.954241   
min         2.000000     18.000000      0.000000       2.541026   
25%     29973.000000     39.000000      0.000000      79.078715   
50%     60572.000000     54.000000      0.000000      96.398561   
75%     90788.000000     65.000000      1.000000     114.310169   
max    121200.000000    103.000000      6.000000     671.298147   

       DaysbetweenDeathandoperation  Preoptransfusionwithin30days  \
count                   5595.000000                  90785.000000   
mean                     476.782127                      0.042672   
std                      421.866740                      0.405840   
min                        0.000000                      0.000000   
25%                      121.000000                      0.000000   
50%                      355.000000                      0.000000   
75%                      746.000000                      0.000000   
max                     1783.000000                     21.000000   

            Intraop  Postopwithin30days  Transfusionintraandpostop  
count  90785.000000        90785.000000               90785.000000  
mean       0.056276            0.018946                   0.075222  
std        0.230455            0.260728                   0.395623  
min        0.000000            0.000000                   0.000000  
25%        0.000000            0.000000                   0.000000  
50%        0.000000            0.000000                   0.000000  
75%        0.000000            0.000000                   0.000000  
max        1.000000           23.000000                  24.000000

In [3]:
# Check for duplicates
# Selecting duplicate rows except first occurrence based on all columns
duplicate = df[df.duplicated()]
 
print("Duplicate Rows :")
 
# Print the resultant Dataframe
duplicate

Duplicate Rows :


Empty DataFrame
Columns: [Indexno, Age, Gender, RCRI score, Anemia category, PreopEGFRMDRD, GradeofKidneydisease, DaysbetweenDeathandoperation, @30daymortality, Preoptransfusionwithin30days, Intraop, Postopwithin30days, Transfusionintraandpostop, AnaestypeCategory, PriorityCategory, TransfusionIntraandpostopCategory, AGEcategory, AGEcategoryOriginal, Mortality, thirtydaymortality, SurgRiskCategory, RaceCategory, CVARCRICategory, IHDRCRICategory, CHFRCRICategory, DMinsulinRCRICategory, CreatinineRCRICategory, GradeofKidneyCategory, Anemiacategorybinned, RDW15.7, ASAcategorybinned, ICUAdmgt24h, Gender_Female, Gender_Male, Gender_NA, Anemia category_Mild, Anemia category_Moderate, Anemia category_None, Anemia category_Severe, Anemia category_NA, GradeofKidneydisease_G1, GradeofKidneydisease_G2, GradeofKidneydisease_G3a, GradeofKidneydisease_G3b, GradeofKidneydisease_G4, GradeofKidneydisease_G5, GradeofKidneydisease_NA, AnaestypeCategory_GA, AnaestypeCategory_RA, AnaestypeCategory_NA, PriorityCategory_Elective, PriorityCategory_Emergency, PriorityCategory_NA, TransfusionIntraandpostopCategory_0 units, TransfusionIntraandpostopCategory_1 unit, TransfusionIntraandpostopCategory_2 or more units, TransfusionIntraandpostopCategory_NA, AGEcategory_18-29, AGEcategory_30-49, AGEcategory_50-64, AGEcategory_65-74, AGEcategory_75-84, AGEcategory_>=85, AGEcategory_NA, AGEcategoryOriginal_18-29, AGEcategoryOriginal_30-49, AGEcategoryOriginal_50-69, AGEcategoryOriginal_>=70, AGEcategoryOriginal_NA, SurgRiskCategory_High, SurgRiskCategory_Low, SurgRiskCategory_Moderate, SurgRiskCategory_NA, RaceCategory_Chinese, RaceCategory_Indian, RaceCategory_Malay, RaceCategory_Others, RaceCategory_NA, GradeofKidneyCategory_G1, GradeofKidneyCategory_G2, GradeofKidneyCategory_G3, GradeofKidneyCategory_G4-G5, GradeofKidneyCategory_NA, Anemiacategorybinned_Mild, Anemiacategorybinned_Moderate/Severe, Anemiacategorybinned_NA, RDW15.7_<= 15.7, RDW15.7_>15.7, RDW15.7_NA, ASAcategorybinned_I, ASAcategorybinned_II, ASAcategorybinned_III, ASAcategorybinned_IV-VI, ASAcategorybinned_NA]
Index: []

[0 rows x 94 columns]

This shows that there are no duplicate rows in the data.

Next, we check for missing values:

In [5]:
# Check number of missing values in each column
missing_values = df.isna().sum()
print(missing_values)

Indexno                        0
Age                            0
Gender                         0
RCRI score                 27424
Anemia category            66916
                           ...  
ASAcategorybinned_I            0
ASAcategorybinned_II           0
ASAcategorybinned_III          0
ASAcategorybinned_IV-VI        0
ASAcategorybinned_NA           0
Length: 94, dtype: int64


In [6]:
# Convert the result to a DataFrame for better visualization
missing_values_df = missing_values.reset_index()
missing_values_df.columns = ['Column', 'Missing Values']

display(missing_values_df)

Column  Missing Values
0                   Indexno               0
1                       Age               0
2                    Gender               0
3                RCRI score           27424
4           Anemia category           66916
..                      ...             ...
89      ASAcategorybinned_I               0
90     ASAcategorybinned_II               0
91    ASAcategorybinned_III               0
92  ASAcategorybinned_IV-VI               0
93     ASAcategorybinned_NA               0

[94 rows x 2 columns]

In [7]:
# Convert the result to a DataFrame for better visualization
missing_values_df = missing_values.reset_index()
missing_values_df.columns = ['Column', 'Missing Values']

# Export to Excel
missing_values_df.to_excel('missing_values.xlsx', index=False)

From the ``missing_values_df`` DataFrame, excluding the outcome variables and the categorical variables which have been one-hot encoded during data cleaning, these columns still have missing values:
- ``RCRI score``: Revised Cardiac Risk Index (float64)
- ``PreopEGFRMDRD``: Estimated glomerular filtration rate (float64)
- ``CVARCRICategory``: Presence of cerebrovascular accident (Boolean)
- ``IHDRCRICategory``: Presence of ischaemic heart disease (Boolean)
- ``CHFRCRICategory``: Presence of congestive heart failure (Boolean)
- ``DMinsulinRCRICategory``: Presence of Diabetes Mellitus (Boolean)
- ``CreatinineRCRICategory``: Presence of Creatinine (Boolean)

In the next section, we shall try different data imputation methods to fill in the missing data.

# Data Imputation using Mean

In this section, we impute the missing values in the ``RCRI score`` column and ``PreopEGFRMDRD`` column with the **mean** value of the respective columns.

Note that data imputation using Mean can be done for the numerical columns only.

In [4]:
df_mean = df.copy()

df_mean['RCRI score'].fillna(df_mean['RCRI score'].mean(), inplace = True)
df_mean['PreopEGFRMDRD'].fillna(df_mean['PreopEGFRMDRD'].mean(), inplace = True)

In [5]:
#Check that there are no more missing values in these two columns:
print(df_mean['RCRI score'].isna().sum())
print(df_mean['PreopEGFRMDRD'].isna().sum())

0
0


In [6]:
df_mean.describe()

Indexno           Age    RCRI score  PreopEGFRMDRD  \
count   90785.000000  90785.000000  90785.000000   90785.000000   
mean    60334.001057     52.253225      0.322296      96.426155   
std     34930.467992     17.087307      0.530313      31.864673   
min         2.000000     18.000000      0.000000       2.541026   
25%     29973.000000     39.000000      0.000000      81.773853   
50%     60572.000000     54.000000      0.000000      96.426155   
75%     90788.000000     65.000000      0.322296     111.439680   
max    121200.000000    103.000000      6.000000     671.298147   

       DaysbetweenDeathandoperation  Preoptransfusionwithin30days  \
count                   5595.000000                  90785.000000   
mean                     476.782127                      0.042672   
std                      421.866740                      0.405840   
min                        0.000000                      0.000000   
25%                      121.000000                      0.000000   
50%                      355.000000                      0.000000   
75%                      746.000000                      0.000000   
max                     1783.000000                     21.000000   

            Intraop  Postopwithin30days  Transfusionintraandpostop  
count  90785.000000        90785.000000               90785.000000  
mean       0.056276            0.018946                   0.075222  
std        0.230455            0.260728                   0.395623  
min        0.000000            0.000000                   0.000000  
25%        0.000000            0.000000                   0.000000  
50%        0.000000            0.000000                   0.000000  
75%        0.000000            0.000000                   0.000000  
max        1.000000           23.000000                  24.000000

# Data Imputation using Median

In this section, we impute the missing values in the ``RCRI score`` column and ``PreopEGFRMDRD`` column with the **median** value of the respective columns.

Similar to Mean data imputation, this can be done for the numerical columns only.

In [7]:
df_median = df.copy()

median_RCRI = df_median['RCRI score'].median()
df_median['RCRI score'].fillna(median_RCRI, inplace = True)

median_PreopEGFRMDRD = df_median['PreopEGFRMDRD'].median()
df_median['PreopEGFRMDRD'].fillna(median_PreopEGFRMDRD, inplace = True)

In [8]:
df_median.describe()

Indexno           Age    RCRI score  PreopEGFRMDRD  \
count   90785.000000  90785.000000  90785.000000   90785.000000   
mean    60334.001057     52.253225      0.224938      96.422863   
std     34930.467992     17.087307      0.550574      31.864675   
min         2.000000     18.000000      0.000000       2.541026   
25%     29973.000000     39.000000      0.000000      81.773853   
50%     60572.000000     54.000000      0.000000      96.398561   
75%     90788.000000     65.000000      0.000000     111.439680   
max    121200.000000    103.000000      6.000000     671.298147   

       DaysbetweenDeathandoperation  Preoptransfusionwithin30days  \
count                   5595.000000                  90785.000000   
mean                     476.782127                      0.042672   
std                      421.866740                      0.405840   
min                        0.000000                      0.000000   
25%                      121.000000                      0.000000   
50%                      355.000000                      0.000000   
75%                      746.000000                      0.000000   
max                     1783.000000                     21.000000   

            Intraop  Postopwithin30days  Transfusionintraandpostop  
count  90785.000000        90785.000000               90785.000000  
mean       0.056276            0.018946                   0.075222  
std        0.230455            0.260728                   0.395623  
min        0.000000            0.000000                   0.000000  
25%        0.000000            0.000000                   0.000000  
50%        0.000000            0.000000                   0.000000  
75%        0.000000            0.000000                   0.000000  
max        1.000000           23.000000                  24.000000

Using Median data imputation, we observe that:

- The mean of **RCRI score** decreased (from 0.322 to 0.225) after replacing missing values in the ``RCRI score`` column with the median.
- The mean of **PreopEGFRMDRD** decreased slightly (from 96.426 to 96.423) after replacing missing values in the ``PreopEGFRMDRD`` column with the median.

The advantage of using Mean data imputation or Median data imputation is that both methods are easy to understand and apply.

However, it does not take into account the variability within the data.

# Linear Interpolation imputation

In this section, we impute the missing values in the ``RCRI score`` column and ``PreopEGFRMDRD`` column using linear interpolation, with Backward fill to handle missing values at the end of the DataFrame, and Forward fill imputations to handle missing values at the start of the DataFrame.

Linear interpolation imputation can be done for the numerical columns only.

In [9]:
df_interpolation = df.copy()

In [10]:
# Interpolation imputation for RCRI score and PreopEGFRMDRD
df_interpolation['RCRI score'].interpolate(method='linear', inplace=True)
df_interpolation['PreopEGFRMDRD'].interpolate(method='linear', inplace=True)

# Forward fill to handle missing values at the start
df_interpolation['RCRI score'].fillna(method='ffill', inplace=True)
df_interpolation['PreopEGFRMDRD'].fillna(method='ffill', inplace=True)

# Backward fill to handle missing values at the end
df_interpolation['RCRI score'].fillna(method='bfill', inplace=True)
df_interpolation['PreopEGFRMDRD'].fillna(method='bfill', inplace=True)


C:\Users\peish\AppData\Local\Temp\ipykernel_3220\3378074887.py:6: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_interpolation['RCRI score'].fillna(method='ffill', inplace=True)
C:\Users\peish\AppData\Local\Temp\ipykernel_3220\3378074887.py:7: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_interpolation['PreopEGFRMDRD'].fillna(method='ffill', inplace=True)
C:\Users\peish\AppData\Local\Temp\ipykernel_3220\3378074887.py:10: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_interpolation['RCRI score'].fillna(method='bfill', inplace=True)
C:\Users\peish\AppData\Local\Temp\ipykernel_3220\3378074887.py:11: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill(

In [11]:
#Check that there are no more missing values in these two columns:
print(df_interpolation['RCRI score'].isna().sum())
print(df_interpolation['PreopEGFRMDRD'].isna().sum())

0
0


In [13]:
df_interpolation.describe()

Indexno           Age    RCRI score  PreopEGFRMDRD  \
count   90785.000000  90785.000000  90785.000000   90785.000000   
mean    60334.001057     52.253225      0.315999      97.787768   
std     34930.467992     17.087307      0.587262      33.144519   
min         2.000000     18.000000      0.000000       2.541026   
25%     29973.000000     39.000000      0.000000      80.655950   
50%     60572.000000     54.000000      0.000000      98.016577   
75%     90788.000000     65.000000      0.500000     115.528070   
max    121200.000000    103.000000      6.000000     671.298147   

       DaysbetweenDeathandoperation  Preoptransfusionwithin30days  \
count                   5595.000000                  90785.000000   
mean                     476.782127                      0.042672   
std                      421.866740                      0.405840   
min                        0.000000                      0.000000   
25%                      121.000000                      0.000000   
50%                      355.000000                      0.000000   
75%                      746.000000                      0.000000   
max                     1783.000000                     21.000000   

            Intraop  Postopwithin30days  Transfusionintraandpostop  
count  90785.000000        90785.000000               90785.000000  
mean       0.056276            0.018946                   0.075222  
std        0.230455            0.260728                   0.395623  
min        0.000000            0.000000                   0.000000  
25%        0.000000            0.000000                   0.000000  
50%        0.000000            0.000000                   0.000000  
75%        0.000000            0.000000                   0.000000  
max        1.000000           23.000000                  24.000000

Using linear interpolation, forward fill and backward fill to impute missing values, we observe that:

- The mean of the **RCRI score** decreased slightly from 0.322 to 0.316

- The mean of the **PreopEGFRMDRD** increased slightly from 96.426 to 97.788

# MICE imputation

In this section, since MICE imputation can be done on both continuous and categorical variables, we conduct MICE imputation on the following columns:

- Continuous variables: ``RCRI score``, ``PreopEGFRMDRD``
- Boolean (binary) variables: ``CVARCRICategory``, ``IHDRCRICategory``, ``CHFRCRICategory``, ``DMinsulinRCRICategory``, ``CreatinineRCRICategory``

In [17]:
df_mice = df.copy()

In [18]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

# Using MICE for imputation
mice_imputer = IterativeImputer()
columns = ['RCRI score', 'PreopEGFRMDRD','CVARCRICategory', 'IHDRCRICategory', 'CHFRCRICategory',
           'DMinsulinRCRICategory', 'CreatinineRCRICategory']
bool_columns = ['CVARCRICategory', 'IHDRCRICategory', 'CHFRCRICategory', 'DMinsulinRCRICategory', 'CreatinineRCRICategory']
imputed_data = mice_imputer.fit_transform(df_mice[columns])

# Convert imputed_data back to a DataFrame
imputed_df = pd.DataFrame(imputed_data, columns = columns)

# Convert the values in 'CVARCRICategory', 'IHDRCRICategory', 'CHFRCRICategory','DMinsulinRCRICategory', 
# 'CreatinineRCRICategory' columns back to Boolean data type
imputed_df[bool_columns] = imputed_df[bool_columns].round().astype(bool)


C:\Users\peish\anaconda3\lib\site-packages\sklearn\utils\validation.py:623: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():


In [19]:
# Replace NaN values with computed values
df_mice[columns] = imputed_df[columns]

imputed_df

RCRI score  PreopEGFRMDRD  CVARCRICategory  IHDRCRICategory  \
0        0.269704      99.710859            False            False   
1        0.322534      96.835026            False            False   
2        0.000838     152.538570            False            False   
3        0.231929     117.231496            False            False   
4        0.000000      98.651255            False            False   
...           ...            ...              ...              ...   
90780    2.000000      48.339582            False            False   
90781    1.000000     126.592489            False            False   
90782    0.383298      86.306771            False            False   
90783    0.000000     101.079879            False            False   
90784    0.000000     125.902498            False            False   

       CHFRCRICategory  DMinsulinRCRICategory  CreatinineRCRICategory  
0                False                  False                   False  
1                False                  False                   False  
2                False                  False                   False  
3                False                  False                   False  
4                False                  False                   False  
...                ...                    ...                     ...  
90780            False                   True                   False  
90781            False                   True                   False  
90782            False                  False                   False  
90783            False                  False                   False  
90784            False                  False                   False  

[90785 rows x 7 columns]

In [20]:
#Check that there are no more missing values in these columns:
print(df_mice['RCRI score'].isna().sum())
print(df_mice['PreopEGFRMDRD'].isna().sum())
print(df_mice['CVARCRICategory'].isna().sum())
print(df_mice['IHDRCRICategory'].isna().sum())
print(df_mice['CHFRCRICategory'].isna().sum())
print(df_mice['DMinsulinRCRICategory'].isna().sum())
print(df_mice['CreatinineRCRICategory'].isna().sum())

0
0
0
0
0
0
0


In [21]:
df_mice.describe()

Indexno           Age    RCRI score  PreopEGFRMDRD  \
count   90785.000000  90785.000000  90785.000000   90785.000000   
mean    60334.001057     52.253225      0.322534      96.835026   
std     34930.467992     17.087307      0.549677      31.914679   
min         2.000000     18.000000     -1.060444       2.541026   
25%     29973.000000     39.000000      0.000000      81.698784   
50%     60572.000000     54.000000      0.000000      98.567073   
75%     90788.000000     65.000000      0.347676     111.439680   
max    121200.000000    103.000000      6.000000     671.298147   

       DaysbetweenDeathandoperation  Preoptransfusionwithin30days  \
count                   5595.000000                  90785.000000   
mean                     476.782127                      0.042672   
std                      421.866740                      0.405840   
min                        0.000000                      0.000000   
25%                      121.000000                      0.000000   
50%                      355.000000                      0.000000   
75%                      746.000000                      0.000000   
max                     1783.000000                     21.000000   

            Intraop  Postopwithin30days  Transfusionintraandpostop  
count  90785.000000        90785.000000               90785.000000  
mean       0.056276            0.018946                   0.075222  
std        0.230455            0.260728                   0.395623  
min        0.000000            0.000000                   0.000000  
25%        0.000000            0.000000                   0.000000  
50%        0.000000            0.000000                   0.000000  
75%        0.000000            0.000000                   0.000000  
max        1.000000           23.000000                  24.000000

Using MICE imputation, we observe that:

- The mean of the **RCRI score** remained the same from 0.322 to 0.322

- The mean of the **PreopEGFRMDRD** increased slightly from 96.426 to 96.835

Comparing the number of TRUE and FALSE values in the other columns before and after MICE data imputation:

In [23]:
print('CVARCRICategory column')
print("\nBefore MICE imputation")
print(df['CVARCRICategory'].value_counts())
print("\nAfter MICE imputation")
print(df_mice['CVARCRICategory'].value_counts())

print('\nIHDRCRICategory column')
print("\nBefore MICE imputation")
print(df['IHDRCRICategory'].value_counts())
print("\nAfter MICE imputation")
print(df_mice['IHDRCRICategory'].value_counts())

print('\nCHFRCRICategory column')
print("\nBefore MICE imputation")
print(df['CHFRCRICategory'].value_counts())
print("\nAfter MICE imputation")
print(df_mice['CHFRCRICategory'].value_counts())

print('\nDMinsulinRCRICategory column')
print("\nBefore MICE imputation")
print(df['DMinsulinRCRICategory'].value_counts())
print("\nAfter MICE imputation")
print(df_mice['DMinsulinRCRICategory'].value_counts())

print('\nCreatinineRCRICategory column')
print("\nBefore MICE imputation")
print(df['CreatinineRCRICategory'].value_counts())
print("\nAfter MICE imputation")
print(df_mice['CreatinineRCRICategory'].value_counts())



CVARCRICategory column

Before MICE imputation
CVARCRICategory
False    60917
True      1543
Name: count, dtype: int64

After MICE imputation
CVARCRICategory
False    89242
True      1543
Name: count, dtype: int64

IHDRCRICategory column

Before MICE imputation
IHDRCRICategory
False    57968
True      4245
Name: count, dtype: int64

After MICE imputation
IHDRCRICategory
False    86495
True      4290
Name: count, dtype: int64

CHFRCRICategory column

Before MICE imputation
CHFRCRICategory
False    63739
True       787
Name: count, dtype: int64

After MICE imputation
CHFRCRICategory
False    89998
True       787
Name: count, dtype: int64

DMinsulinRCRICategory column

Before MICE imputation
DMinsulinRCRICategory
False    61907
True      2003
Name: count, dtype: int64

After MICE imputation
DMinsulinRCRICategory
False    88782
True      2003
Name: count, dtype: int64

CreatinineRCRICategory column

Before MICE imputation
CreatinineRCRICategory
False    72760
True      2282
Name: count, dt

The advantage of MICE data imputation is that it can impute missing values for both continuous and categorical variables. However, it's more complex to implement compared to simpler methods like Mean or Median imputation.

# Regression imputation

In this section, we first use Mean data imputation to impute the missing values in the ``PreopEGFRMDRD`` column, then use Linear Regression to impute the missing values in the ``RCRI score`` column. 

Linear Regression imputation can be done for the numerical columns only.

In [29]:
df_regression = df.copy()

In [30]:
from sklearn.linear_model import LinearRegression
from sklearn.impute import SimpleImputer


# Step 1: Impute missing values in PreopEGFRMDRD using mean
imputer = SimpleImputer(strategy='mean')
df_regression['PreopEGFRMDRD'] = imputer.fit_transform(df_regression[['PreopEGFRMDRD']])

# Step 2: Regression imputation for RCRI score
regressor_rcri = LinearRegression()
non_na_data_rcri = df_regression.dropna(subset=['RCRI score', 'PreopEGFRMDRD'])
regressor_rcri.fit(non_na_data_rcri[['PreopEGFRMDRD']], non_na_data_rcri['RCRI score'])

missing_rcri = df_regression[df_regression['RCRI score'].isna()]
if not missing_rcri.empty:
    df_regression.loc[df_regression['RCRI score'].isna(), 'RCRI score'] = regressor_rcri.predict(missing_rcri[['PreopEGFRMDRD']])

# Check if there are still missing values in PreopEGFRMDRD after mean imputation
if df_regression['PreopEGFRMDRD'].isna().sum() > 0:
    # Regression imputation for PreopEGFRMDRD
    regressor_preop = LinearRegression()
    non_na_data_preop = df_regression.dropna(subset=['PreopEGFRMDRD', 'RCRI score'])
    regressor_preop.fit(non_na_data_preop[['RCRI score']], non_na_data_preop['PreopEGFRMDRD'])

    missing_preop = df_regression[df_regression['PreopEGFRMDRD'].isna()]
    if not missing_preop.empty:
        df_regression.loc[df_regression['PreopEGFRMDRD'].isna(), 'PreopEGFRMDRD'] = regressor_preop.predict(missing_preop[['RCRI score']])


C:\Users\peish\anaconda3\lib\site-packages\sklearn\utils\validation.py:623: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
C:\Users\peish\anaconda3\lib\site-packages\sklearn\utils\validation.py:623: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
C:\Users\peish\anaconda3\lib\site-packages\sklearn\utils\validation.py:623: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
C:\Users\peish\anaconda3\lib\site-packages\sklearn\utils\validation.py:623: FutureWarning: is_sparse is deprecated and will be removed in a future version

In [31]:
#Check that there are no more missing values in these two columns:
print(df_regression['RCRI score'].isna().sum())
print(df_regression['PreopEGFRMDRD'].isna().sum())

0
0


In [32]:
df_regression.describe()

Indexno           Age    RCRI score  PreopEGFRMDRD  \
count   90785.000000  90785.000000  90785.000000   90785.000000   
mean    60334.001057     52.253225      0.320926      96.426155   
std     34930.467992     17.087307      0.539922      31.864673   
min         2.000000     18.000000     -1.856552       2.541026   
25%     29973.000000     39.000000      0.000000      81.773853   
50%     60572.000000     54.000000      0.000000      96.426155   
75%     90788.000000     65.000000      0.402673     111.439680   
max    121200.000000    103.000000      6.000000     671.298147   

       DaysbetweenDeathandoperation  Preoptransfusionwithin30days  \
count                   5595.000000                  90785.000000   
mean                     476.782127                      0.042672   
std                      421.866740                      0.405840   
min                        0.000000                      0.000000   
25%                      121.000000                      0.000000   
50%                      355.000000                      0.000000   
75%                      746.000000                      0.000000   
max                     1783.000000                     21.000000   

            Intraop  Postopwithin30days  Transfusionintraandpostop  
count  90785.000000        90785.000000               90785.000000  
mean       0.056276            0.018946                   0.075222  
std        0.230455            0.260728                   0.395623  
min        0.000000            0.000000                   0.000000  
25%        0.000000            0.000000                   0.000000  
50%        0.000000            0.000000                   0.000000  
75%        0.000000            0.000000                   0.000000  
max        1.000000           23.000000                  24.000000

Using Regression imputation, we observe that:

- The mean of the **RCRI score** slightly decreased from 0.322 to 0.321

- The mean of the **PreopEGFRMDRD** remained the same from 96.426 to 96.426

Although Regression imputation uses the relationships between variables which may lead to more accurate imputations than simple methods like Mean or Median imputation, it assumes that the relationships between the variables is linear, which might not be actually true.

# Random Forest Imputation

In [8]:
df_rf = df.copy()

In [9]:
from sklearn.ensemble import RandomForestRegressor
import pandas as pd

# Step 1: Impute missing values in PreopEGFRMDRD using mean
imputer = SimpleImputer(strategy='mean')
df_rf['PreopEGFRMDRD'] = imputer.fit_transform(df_rf[['PreopEGFRMDRD']])

# Step 2: Random forest imputation for RCRI score
rf_imputer_rcri = RandomForestRegressor(n_estimators=100)
non_na_data_rcri = df_rf.dropna(subset=['RCRI score', 'PreopEGFRMDRD'])
rf_imputer_rcri.fit(non_na_data_rcri[['PreopEGFRMDRD']], non_na_data_rcri['RCRI score'])

missing_rcri = df_rf[df_rf['RCRI score'].isna()]
if not missing_rcri.empty:
    df_rf.loc[df_rf['RCRI score'].isna(), 'RCRI score'] = rf_imputer_rcri.predict(missing_rcri[['PreopEGFRMDRD']])

# Check if there are still missing values in PreopEGFRMDRD after mean imputation
if df_rf['PreopEGFRMDRD'].isna().sum() > 0:
    # Random forest imputation for PreopEGFRMDRD
    rf_imputer_preop = RandomForestRegressor(n_estimators=100)
    non_na_data_preop = df_rf.dropna(subset=['PreopEGFRMDRD', 'RCRI score'])
    rf_imputer_preop.fit(non_na_data_preop[['RCRI score']], non_na_data_preop['PreopEGFRMDRD'])

    missing_preop = df_rf[df_rf['PreopEGFRMDRD'].isna()]
    if not missing_preop.empty:
        df_rf.loc[df_rf['PreopEGFRMDRD'].isna(), 'PreopEGFRMDRD'] = rf_imputer_preop.predict(missing_preop[['RCRI score']])

C:\Users\peish\anaconda3\lib\site-packages\sklearn\utils\validation.py:623: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
C:\Users\peish\anaconda3\lib\site-packages\sklearn\utils\validation.py:623: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
C:\Users\peish\anaconda3\lib\site-packages\sklearn\utils\validation.py:623: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
C:\Users\peish\anaconda3\lib\site-packages\sklearn\utils\validation.py:623: FutureWarning: is_sparse is deprecated and will be removed in a future version

In [10]:
#Check that there are no more missing values in these two columns:
print(df_rf['RCRI score'].isna().sum())
print(df_rf['PreopEGFRMDRD'].isna().sum())

0
0


In [11]:
df_rf.describe()

Indexno           Age    RCRI score  PreopEGFRMDRD  \
count   90785.000000  90785.000000  90785.000000   90785.000000   
mean    60334.001057     52.253225      0.318098      96.426155   
std     34930.467992     17.087307      0.576087      31.864673   
min         2.000000     18.000000      0.000000       2.541026   
25%     29973.000000     39.000000      0.000000      81.773853   
50%     60572.000000     54.000000      0.000000      96.426155   
75%     90788.000000     65.000000      0.342938     111.439680   
max    121200.000000    103.000000      6.000000     671.298147   

       DaysbetweenDeathandoperation  Preoptransfusionwithin30days  \
count                   5595.000000                  90785.000000   
mean                     476.782127                      0.042672   
std                      421.866740                      0.405840   
min                        0.000000                      0.000000   
25%                      121.000000                      0.000000   
50%                      355.000000                      0.000000   
75%                      746.000000                      0.000000   
max                     1783.000000                     21.000000   

            Intraop  Postopwithin30days  Transfusionintraandpostop  
count  90785.000000        90785.000000               90785.000000  
mean       0.056276            0.018946                   0.075222  
std        0.230455            0.260728                   0.395623  
min        0.000000            0.000000                   0.000000  
25%        0.000000            0.000000                   0.000000  
50%        0.000000            0.000000                   0.000000  
75%        0.000000            0.000000                   0.000000  
max        1.000000           23.000000                  24.000000

Using Random Forest imputation, we observe that:

- The mean of the **RCRI score** slightly decreased from 0.322 to 0.318

- The mean of the **PreopEGFRMDRD** remained the same from 96.426 to 96.426